src/Common/src/Interop/Unix/libc/Interop.strerror.cs

```diff
@@ -10,15 +10,70 @@ internal static partial class Interop
 {
     internal static partial class libc
     {
-        internal unsafe static string strerror(int errno) // thread-safe version of strerror that internally uses strerror_r, which is thread-safe
+        internal static string strerror(int errno)
         {
-            const int bufferLength = 1024; // length long enough for most any Unix error messages
-            byte* buffer = stackalloc byte[bufferLength];
-            IntPtr errorPtr = (IntPtr)strerror_r(errno, buffer, (IntPtr)bufferLength);
-            return Marshal.PtrToStringAnsi(errorPtr); // TODO: Use the proper Encoding
+            string result = s_isGnu.Value ? strerror_gnu(errno) : strerror_xsi(errno);
+            // System.Diagnostics.Debug.WriteLine("strerror: " + result + "\n" + System.Environment.StackTrace); // uncomment to aid debugging
+            return result;
         }
 
+        // strerror is not thread-safe; instead, there is the strerror_r function, which is thread-safe.
+        // However, there are two versions of strerror_r:
+        // - GNU: char* strerror_r(int, char*, size_t);
+        // - XSI: int   strerror_r(int, char*, size_T);
+        // The former may or may not use the supplied buffer, and returns the error message string.
+        // The latter stores the error message string into the supplied buffer.
+        // Due to the different return values, we don't want to just choose one signature
+        // and try to use a heuristic to deduce which version we're dealing with, as that
+        // could result in a stack imbalance due to the varied return result size on 64-bit.  Instead,
+        // we detect whether we're compiled against GNU's libc by trying to invoke its
+        // version-getting method, and invoke the right signature based on that.
+
+        private static Lazy<bool> s_isGnu = new Lazy<bool>(() =>
+        {
+            try
+            {
+                // Just try to call the P/Invoke.  If it succeeds, this is GNU libc.
+                gnu_get_libc_version();
+                return true;
+            }
+            catch
+            {
+                // Otherwise, it's not.
+                return false;
+            }
+        });
+
+        private const int MaxErrorMessageLength = 1024; // length long enough for most any Unix error messages
+
+        private static unsafe string strerror_gnu(int errno)
+        {
+            byte* buffer = stackalloc byte[MaxErrorMessageLength];
```

Comment:
Why is that better?  This way there are no managed allocations for the buffer.  For the StringBuilder approach it would still require allocating the StringBuffer, setting its Capacity appropriately, etc.